psi4/src/psi4/libmints/mintshelper.cc

```diff
@@ -1812,188 +1812,13 @@ SharedMatrix MintsHelper::perturb_grad(SharedMatrix D) {
-                    }
-                }
-
-                // Qx
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][0] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Qy
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][1] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Qz
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][2] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-            } else {
-                // Ylambda is zero, so we just advance the pointer to the buffer
-                ref += 6 * nP * nQ;
-            }
-
-            /*
-             * Mu Z derivatives
-             */
-            if (zlambda != 0.0) {
-                prefac = perm * zlambda;
-                // Px
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aP][0] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Py
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aP][1] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Pz
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aP][2] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Qx
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][0] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Qy
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][1] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-
-                // Qz
-                for (int p = 0; p < nP; p++) {
-                    for (int q = 0; q < nQ; q++) {
-                        Pp[aQ][2] += prefac * Dp[p + oP][q + oQ] * (*ref++);
-                    }
-                }
-            }
-        }
-    }
-    return ret;
+    int natoms = basisset_->molecule()->natom();
+    auto perturbation_gradient = std::make_shared<Matrix>("Perturbation Gradient", natoms, 3);
+    auto dipole_gradients = dipole_grad(D);
+    double lambdas[3] = {xlambda, ylambda, zlambda};
+    C_DGEMM('n', 't', 3 * natoms, 1, 3, 1.0, dipole_gradients->pointer()[0], 3, &lambdas[0], 3, 0.0,
```

Comment:
Well, that's easier.

Reply:
Sure is.  The pert_grad routine computed this contraction within the loops, so wasn't suitable for the dipole gradients.  By computing the uncontracted dipole gradients and GEMMing them, we avoid repetition as you suggested - good call.

Reply:
Good call pulling the lambdas out of the loops.